RAS2FIM Implementation

CoreyKrewson-NOAA opened the pull request:
This PR will look to implement ras2fim modeling capabilities into the live and static FIM services. After implemented, the flood services will show ras2fim where we have ras2fim data and streamflows are within valid ras2fim ranges. If ras2fim is not available or can't be used, HAND will then be used.

CoreyKrewson-NOAA committed:
added ras2fim insertion logic into fim_data_prep lambda

CoreyKrewson-NOAA committed:
updated data sql files to only get feature id and streamflow

CoreyKrewson-NOAA committed:
updated schema and table for ras2fim geocurves

CoreyKrewson-NOAA committed:
updated mapx fields to new fields for ras2fim

TylerSchrag-NOAA commented on Core/LAMBDA/viz_functions/viz_fim_data_prep/data_sql/ana_inundation_hi.sql:
I'm not sure if this really matters, but I'm wondering if we should just put 'rf.high_water_threshold > 0' in the where clause of all of these to filter out the error codes and weird negative recurrence flow values that exist for some features in the conus table?

CoreyKrewson-NOAA replied:
yeah we could do that. This example is for HI though. For CONUS, wouldn't the existing max_forecast.maxflow_1hour_cfs >= rf.high_water_threshold conditions work? If 0 is the default value then we would never get any negative value reaches

TylerSchrag-NOAA replied:
Yeah, I just commented on the HI one, but was thinking about all of the configs. I think it's likely I'm mis-remembering how these get filtered out somewhere else, but I'm referring to all the reaches in the conus recurrence flows table with a high_water_threshold of -21474836.48.

![image](https://github.com/NOAA-OWP/hydrovis/assets/97621365/0b0c6537-82ea-4ac0-8d35-45a5b252923a)

CoreyKrewson-NOAA replied:
oh duh, sorry my comment didn't even make sense. Yeah I am with you. We should probably exclude those

TylerSchrag-NOAA commented on Core/LAMBDA/viz_functions/viz_fim_data_prep/lambda_function.py:
While I like the ingenuity of inserting a dynamic WITH clause (especially as we have a growing list of configs requring new sql files)... I'm not wild about the amount of SQL living inline in the python code on this. We lose some amount of simple testability doing it that way, where one can't just drop the SQL from the SQL file straight into pgadmin like everything else (not that it's terribly hard to copy and paste together). 

This is nitpicky, but I also don't love that the streamflow_sql gets ran twice back-to-back - gotta be a way for us to optimize that.

I think it's fine to leave this as-is for now... but I'd like to think about how we might reengineer this a bit more as we move towards implementing the flood stacks at scale. I gotta think there is some way for us to augment the max_flows step to make this part simpler.

CoreyKrewson-NOAA replied:
I think without the WITH clause though, there are so many lines that are duplicated and it would be easy to miss something. This way just boils it down to the simplest unique sql for each run. Maybe a middle solution is to extract the sql in the python and create a fim_template.sql file. That way it will be easier to copy/paste and people don't have to dig while maintaining the simplicity of the config sql with what it actually wants to process.

yeah I didn't love that as well.

TylerSchrag-NOAA replied:
Nice adjustment for now! As we've discussed, I've always been torn on the best approach to dynamic SQL vs. the simplicity, yet redundant code, of having definitive SQL files. A designated template sql file is a good idea... I bet we could use that in some other spots as well.

TylerSchrag-NOAA commented on Core/LAMBDA/viz_functions/viz_fim_data_prep/lambda_function.py:
```diff
@@ -251,3 +258,133 @@ def write_data_csv_file(product, fim_config, huc, reference_date, huc_data):
+                lag(discharge_cms) OVER (PARTITION BY feature_id ORDER BY discharge_cms) as previous_discharge_cms,
+                stage_ft,
+                version
+            FROM ras2fim.geocurves
```
Make sure to try adding a feature_id index to the geocurves table... that might speed this up a bit more.

I'd also suggest adding previous_discharge_cms as a permanent column to the geocurves table, and testing a two-column primary key of feature_id/previous_discharge_cms.

CoreyKrewson-NOAA replied:
yeah good idea on just permanently adding the previous discharge column. Limit the dynamic SQL as much as we can

TylerSchrag-NOAA commented on Core/LAMBDA/viz_functions/viz_publish_service/services/analysis_assim/ana_inundation_extent.mapx:
```diff
@@ -283,7 +283,7 @@
             "type" : "CIMFieldDescription",
             "alias" : "NWM Feature ID",
-            "fieldName" : "feature_id",
+            "fieldName" : "nwm_feature_id",
```
Are we planning to change feature_id to nwm_feature_id everywhere? I'd suggest we keep that consistent across the whole stack if so and either change it everywhere if so, since it's our main primary key.

CoreyKrewson-NOAA replied:
yeah good point. Let's revert the feature id. Originally I was trying to think of a way to really distinguish between what the IDs are for the FIM and what the IDs are for the streamflow model so I was just more explicit. Maybe instead of Hydro ID, we use FIM ID?

TylerSchrag-NOAA replied:
Yeah, I like where you're going. It's a big change, so can happen whenever the timing is best, but nwm_feature_id is better than feature_id, and fim_hydro_id is better than hydro_id. Maybe nws_lid instead of the various aliases for that one as well.

CoreyKrewson-NOAA replied:
Let's hold off for now because it is a big change like you said. With the public FIM being pretty close I am hesitant for a big change.

TylerSchrag-NOAA reviewed (dismissed):
Few comments, that can wait if you want. I don't see anything else.

CoreyKrewson-NOAA committed:
moved sql from python to template files
added where clause to all fim_data sql to not process high water threshold 0 and below

CoreyKrewson-NOAA committed:
updated all hand_stage instances to fim_stage
